### PR TITLE
Add instrumentation for HTTP client

### LINKED ISSUE #202: Add instrumentation for the http client embedded in the k8s client for kubearchive rest api server
We need logs for the requests done from the api server to the kubernetes API.

That imply configuring the kubernetes client with the proper [Transport](https://github.com/kubernetes/client-go/blob/master/rest/config.go#L104)

I think that we can follow [this piece of code](https://uptrace.dev/get/instrument/opentelemetry-net-http.html#instrumenting-http-client) to add the otel transport for instrumenting the http client of the k8s client

### PR DESCRIPTION
**Pull request by rh-hemartin**
## Which Issue(s) this Pull Request Fixes

Resolves #202 

## Release Note

```release-note
NONE
```

## Notes for Reviewers

* The approach for testing (creating a RESTClient) has been inspired by code that tests the kubectl code. Its bulky, but gets the work done. I didn't look at it enough to see how to make it easier, if that is possible.

### CODE REVIEW

**Comment by rh-hemartin** [comment]
I'm having some problems related with testing. Due to the implementation of the fake library, I can't access the RESTClient on the Discovery field of the fake Clientset.

File: cmd/api/discovery/apiresources_test.go [hunk 1]
```diff
@@ -33,7 +42,33 @@ func TestGetAPIResource(t *testing.T) {
...
+				default:
+					t.Fatalf("unexpected request: %#v\n%#v", req.URL, req)
+					return nil, nil
+				}
+			default:
+				t.Fatalf("unexpected request: %s %#v\n%#v", req.Method, req.URL, req)
+				return nil, nil
+			}
+		}),
+	}
```
Comment (maruiz93) [inline cmd/api/discovery/apiresources_test.go:71]: I also played around with how to mock the restClient and I came out with a different solution.
I wanted to take advantage of the faked `ServerResourcesForGroupVersion` that was being used before, so I call it and mock the RestClient after that:

```suggestion
           // call ServerResourcesForGroupVersion and properly set the response and the status of the server
			apiResources, err := fakeDiscovery.ServerResourcesForGroupVersion(fmt.Sprintf("%s/%s", tc.group, tc.version))
			status := http.StatusOK
			response := ""
			if err != nil {
				response = err.Error()
				status = http.StatusBadRequest
			} else {
				buf := bytes.NewBufferString("")
				serializer.Encode(apiResources, buf)
				response = buf.String()
			}
           // mock the restClient with the given status and response
			restClient := &fakerest.RESTClient{
				Client: fakerest.CreateHTTPClient(func(request *http.Request) (*http.Response, error) {
					resp := &http.Response{
						StatusCode: status,
						Body: io.NopCloser(
							strings.NewReader(response)),
					}
					return resp, nil
				}),
				NegotiatedSerializer: scheme.Codecs.WithoutConversion(),
				GroupVersion:         schema.GroupVersion{Version: "v1"},
				VersionedAPIPath:     fmt.Sprintf("/apis/%s/%s", tc.group, tc.version),
			}
```

For that to work is needed also the serializer:

```
	s := runtime.NewScheme()
	serializer := json.NewSerializerWithOptions(json.DefaultMetaFactory, s, s, json.SerializerOptions{
		Pretty: true,
		Yaml:   false,
		Strict: true,
	})
```

What I like of this solution is that the logic is mostly deferred to the fake module, I just put in place the glue between the faked `ServerResourcesForGroupVersion` and the RestClient

**Review by maruiz93 — CHANGES_REQUESTED** [review]

File: cmd/api/discovery/apiresources.go [hunk 2]
```diff
@@ -5,22 +5,29 @@ import (
...
+			GroupVersion: groupVersion,
+		}
+		// TODO: when we support the core API we need to change this
+		url := fmt.Sprintf("/apis/%s", groupVersion)
+		err := client.Get().AbsPath(url).Do(c.Request.Context()).Into(resources)
 		if err != nil {
-			abort.Abort(c, err.Error(), http.StatusBadRequest)
-			return
+			abort.Abort(c, fmt.Errorf("Unable to retrieve information from '%s', error: %w", url, err).Error(), http.StatusBadRequest)
 		}
```
[the remark this answers is not in the document]
Comment (rh-hemartin) [reply cmd/api/discovery/apiresources.go]: I separated both stages, as you suggested.

File: cmd/api/discovery/apiresources_test.go (hunk 1 above)
Comment (rh-hemartin) [reply cmd/api/discovery/apiresources_test.go]: Good idea! I went with this idea but dropped the serializer in favor of just using `json.Marshal`.

Comment (maruiz93) [reply cmd/api/discovery/apiresources_test.go]: Nice! I like how it looks now :)

File: cmd/api/discovery/apiresources.go [hunk 3]
```diff
@@ -5,22 +5,40 @@ import (
...
-	"k8s.io/client-go/discovery"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/rest"
 
 	"github.com/kubearchive/kubearchive/cmd/api/abort"
 )
 
 // GetAPIResource set apiResource attribute in the context based on the path parameters group version and resourceType
-func GetAPIResource(discoveryInterface discovery.DiscoveryInterface) gin.HandlerFunc {
+func GetAPIResource(client rest.Interface) gin.HandlerFunc {
```
Comment (maruiz93) [inline cmd/api/discovery/apiresources.go:15]: How about adding a comment referencing the issue you opened to get DiscoveryInterface to admit a context?

**Review by maruiz93 — APPROVED** [review]

File: cmd/api/discovery/apiresources.go (hunk 3 above)
Comment (rh-hemartin) [reply cmd/api/discovery/apiresources.go:19]: Added a paragraph for context.

**Review by skoved — APPROVED** [review]

**Pull request merged**